src/Bicep.Core/Workspaces/SourceFileGroupingBuilder.cs

```diff
@@ -135,56 +135,61 @@ private FileResolutionResult PopulateRecursive(Uri fileUri, ModuleReference? mod
             switch (fileResult.File)
             {
                 case BicepFile bicepFile:
-                {
-                    foreach (var restorable in bicepFile.ProgramSyntax.Children.OfType<IForeignTemplateReference>())
                     {
-                        var (childModuleReference, uriResult) = GetModuleRestoreResult(fileUri, restorable);
-
-                        uriResultByModule.GetOrAdd(bicepFile, f => new())[restorable] = uriResult;
-                        if (uriResult.FileUri is null)
+                        foreach (var restorable in bicepFile.ProgramSyntax.Children.OfType<IForeignTemplateReference>())
                         {
-                            continue;
-                        }
+                            if (restorable.GetType() == typeof(ProviderDeclarationSyntax))
+                            {
+                                continue;
+                            }
```

Comment:
It's probably safer to check if `restorable is ProviderDeclarationSyntax` in case anyone subclasses `ProviderDeclarationSyntax` in a future PR.

Btw, if provider declarations are opting out of the cache-aware restore process here, do they need to implement `IForeignTemplateReference`?

Reply:
I ddidn't intend for them to opt-out of the cache-aware restore process, can you elaborate?

Reply:
This is where other `IForeignTemplateReference` statements are checked to see if a copy has already been cached.